contracts/sdk/IpcContract.sol

```diff
@@ -0,0 +1,151 @@
+// SPDX-License-Identifier: MIT OR Apache-2.0
+pragma solidity 0.8.19;
+
+import {IpcEnvelope, ReceiptMsg, IpcMsg, IpcMsgKind} from "../src/structs/CrossNet.sol";
+import {EMPTY_BYTES} from "../src/constants/Constants.sol";
+import {IGateway} from "../src/interfaces/IGateway.sol";
+
+// Interface that needs to be implemented by IPC-enabled contracts.
+// This is really convenient to call it from other contracts.
+interface IfaceIpcContract {
+    /// @notice Entrypoint for cross-net messages in IPC-enabled contracts.
+    // solhint-disable-next-line func-name-mixedcase
+    function IpcEntrypoint(IpcEnvelope calldata crossMsg) external payable returns (bytes memory ret);
+}
+
+abstract contract IpcContract {
+    // list of messages in-flight for which the contract
+    // hasn't received an ACK.
+    mapping(bytes32 => IpcEnvelope) public inFlightMsgs;
+    // auxiliary mapping to track the height at which certain
+    // in-flight messages were received.
+    mapping(uint256 => bytes32[]) public inFlightMsgHeights;
+    // last height where there messages were pruned
+    uint256 private lastGcBlock;
+
+    // The adderss of the gateway in the subnet.
+    address public gatewayAddr;
+    /// Number of blocks that the contract will wait for a receipt
+    /// before garbage collecting the in-flight message.
+    uint256 public timeoutBlocks;
```

Comment:
- A timeout is a simple solution but it's dangerous in the context of distributed transactions.
- IPC as a framework enforces no timeouts, so it's possible for us to drop a message here (and potentially roll back any local actions, e.g. return funds to the user) only for an ACK to be delivered the epoch later. This would throw the contract/dapp into an inconsistent state, with potential double-booking of funds and no way to amend the state.
- Dropping a message on timeout may require a local rollback action, which currently this interface does not support.
- State ballooning, whilst undesirable, only affects long-term gas costs in FEVM. Freeing state does not refund gas in FEVM (unlike the EVM). While it _does_ make the amortized cost of writing lower over time (due to lower HAMT depth), we can regard that purely as an optimization. In fact, we need to store a whole entirely mapping _only_ to do timeout tracking, so I'll go out on a limb and say that timeout tracking may even be counterproductive cost-wise.
- The GC logic, as currently implemented, puts the cost of clearing the map on the first entrypoint call for a non-empty epoch, which is unfair.

Therefore:
1. I am going to take the liberty to remove the timeout machinery here.
2. Instead, I'll add a method for the owner to explicitly _drop_ messages by hash. This allows the owner to run whatever local logic they want if they really care about dropping potentially expired messages.
3. By making this explicit, it also forces the contract author to think about rollback effects that may be associated by declaring a message dead.

Reply:
Solved in #606, merged here.